ui/raidboss/data/04-sb/ultimate/ultima_weapon_ultimate.ts

```diff
@@ -1132,6 +1151,63 @@ const triggerSet: TriggerSet<Data> = {
       },
     },
     // --------- Titan ----------
+    {
+      id: 'UWU Titan Last Move Collector',
+      type: 'ActorMove',
+      netRegex: { id: '4[0-9a-fA-F]{7}', capture: true },
+      condition: (data, matches) =>
+        data.phase === 'titan' &&
+        data.bossId.titan === matches.id,
+      run: (data, matches) => {
+        data.lastTitanMove = matches;
+      },
+    },
+    {
+      id: 'UWU Titan Jump Direction',
+      type: 'NameToggle',
+      netRegex: { id: '4[0-9a-fA-F]{7}', toggle: '00', capture: true },
+      condition: (data, matches) =>
+        data.phase === 'titan' &&
+        data.bossId.titan === matches.id &&
+        data.lastTitanMove !== undefined,
+      // There is sometimes a "late" ActorMove packet that can arrive anywhere from ~45-180ms after
+      // the NameToggle packet. A delay of 0.4s should be enough for safety without a belated call.
+      delaySeconds: 0.4,
+      alertText: (data, _matches, output) => {
+        let safeStr = output.safe!({ dir: output.unknown!() });
+        const lastMove = data.lastTitanMove;
+        if (lastMove === undefined)
+          return safeStr;
+
+        const titanX = parseFloat(lastMove.x);
+        const titanY = parseFloat(lastMove.y);
+        const titanHdg = parseFloat(lastMove.heading);
+
+        // Titan's final heading will be facing one of the four cardinal jump sites.
+        // We can identify the correct site by comparing Titan's heading to the heading of the
+        // jump site (toward Titan) where the difference is +/-pi -- in other words, a
+        // 180-degree angle deviation indicative of an identical ray.
+        for (const site of jumpSites) {
+          const hdgToTitan = getRelativeHdg(titanX, titanY, site.x, site.y);
+          const ray = Math.abs(titanHdg - hdgToTitan);
+          if (ray >= 3 && ray <= 3.28) // should be pi, but allow for rounding errors etc.
+            safeStr = output.safe!({ dir: output[site.safe]!() });
```

Comment:
Our usual practice is to early out whenever possible. While it doesn't really make much of a difference here from a code or performance perspective, I think it still makes sense to `return output.safe...` here.